accepted/2024/release-staging-and-deprecation.md

```diff
@@ -0,0 +1,193 @@
+
+## Motivation
+
+<!-- Why are we doing this? What pain points does this solve? What is the expected outcome? -->
+Many products (Apps / Tools / Frameworks / Utilities / etc. ...) consist of multiple interdependent packages which should be published together as a single set.
+ Currently, NuGet provides no functionality for defining / grouping packages into such sets.
+ This results in package owners having to keep track of these package sets on their own.
+ There are three specific sets of scenarios which this feature will simplify dramatically:
+1. When preparing a release, product owners would like to have a mechanism to pre-stage their product packages,
+ so that the release will consist only of a single action of marking the set of packages as `published`,
+ which would result all the packages in the set to be marked as published at the same time.
+2. Sometimes developers have to take a dependency on several related packages in their projects, which are part of the same "release".
+ Without having a concept of the `release` developers have to guess which versions of those packages are compatible.
+ As a result, most of the time developers would simply pick the latest version of all these packages, which sometimes can lead to incompatibility issues down the road.
+3. Keeping a package healthy on NuGet also means keeping the list of packages aligned with it support policy.
+Hence, package owners try to deprecate packages which are out of support. Without ability to group packages,
+ package owners end up keeping track of all the packages which belong to the same release and run the deprecation over a discrete set of packages when necessary. Being able to group packages together would enable deprecating a whole release at once, knowing that all the packages which belong to that release will get deprecated.
+
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
+A new NuGet CLI command option `stage` will be added to the `push` command, which will allow a user to stage a package to nuget, specifying a group identifier. (dotnet nuget push --stage [group-id])
+ Staged packages will be uploaded to NuGet and will be validated as part of the staging process, so that if a package being staged fails validation, the staging fails. After uploading multiple packages, the user can then call `nuget push --group [group-id]`
+ command to request all the packages from the earlier group to become available on the server as `published`.
+Note, that the `group` information will only be stored on the server and has nothing to do with how a package was built.
+
+Groups are scoped to an owner. So a package published by `Owner 1` into `Group 1` has nothing to do with another package published
```

Comment:
Packages can be owned by multiple people. We need a nice way of fitting "group ownership" into that, too. If only the creator of a group can interact with it (add more packages, publish everything, ...), it's likely going to be the easiest to implement, but might also introduce some confusion when package has several owners.

### Scenario 1
Package1 and Package2 both have 2 owners: Owner1 and Owner2.
If Owner1 stages Package1 should Owner2 be able to stage Package2 to the same group? Will customers expect to be able to do that?

### Scenario 2
Package1 has Owner1 and Owner2 as an owner.
Package2 has Owner2 and Owner3.
Owner1 stages Package1. Should Owner3 be able to stage Package2 to the same group?

Reply:
I may have not worded this nicely, and I will try to update the wording in the proposal, but basically, groups / stages have no concept of an `owner`. The only ownership concept comes from packages. Hence, imagine a scenario where user A and userB are co-owners of a package. Any one of them can stage that package and the other one can handle any operation with the group the package has been associated with. If userB discards the stage, then userA won't be able to find that package in the stage any more.

Let's apply this to your scenarios:

### Scenario 1
Yes, Owner 2 can stage Package 2 to the same group. As long as all the packages in that group are co-owned by both of them, they will see exactly the same thing.

### Scenario 2
Owner 3 can stage Package 2 to a group called with the same name. However, these are two different groups.

Think about all this ownership thing from data model perspective.
The NuGet server will associate `group id` with each `package` that's staged / published. So when asking the question `which packages belong to this group` depending on which packages you have access to you will be getting a subset from the whole list.

Hope this clarifies it.

Reply:
I've just updated the writeup to include all this, so that it's clear going forward.

Reply:
So, Owner2 in this scenario for the same group will see different group content than Owner1 and Owner3?

Reply:
That's correct!
However, if the publishing request is going to result in a different set of packages being published, than the original set that was staged, then the request will fail with a warning, and the user will need to explicitly specify the `--force` option to indicate that they understand what's happening and why.